contrib/coredns/policy/policy.go

```diff
@@ -150,7 +174,47 @@ func (p *PolicyMiddleware) ServeDNS(ctx context.Context, w dns.ResponseWriter, r
 	}
 
 	if response.Permit {
-		return middleware.NextOrFailure(p.Name(), p.Next, ctx, w, r)
+		lr := new(NewLocalResponseWriter)
+		status,err := middleware.NextOrFailure(p.Name(), p.Next, ctx, lr, r)
+		if err != nil {
+			return status, err
+		}
+
+		state := request.Request{W: lr, Req: lr.Msg}
+
+		rr := lr.Msg.Answer[0]
+		switch state.Family() {
+		case 1:
+			attrs = append(attrs, &pb.Attribute{Id: "address", Type: "address", Value: rr.(*dns.A).A.String()})
+		case 2:
+			attrs = append(attrs, &pb.Attribute{Id: "address", Type: "address", Value: rr.(*dns.AAAA).AAAA.String()})
+
+		}
+
+		attrs[0].Value = "response"
+
+		var response Response
+		err = p.pdp.Validate(ctx, pb.Request{Attributes: attrs}, &response)
+
+		if err != nil {
+			log.Printf("[ERROR] Policy validation failed due to error %s\n", err)
+			return dns.RcodeServerFailure, err
+		}
+
+		if response.Permit == true && response.Redirect == nil {
+			w.WriteMsg(r)
+			return status, nil
+		}
+
+		if response.Redirect != nil {
+			log.Printf("[INFO] Redirecting request")
+			return p.redirect(response.Redirect.String(), w, r)
+		}
+
+		if response.Permit == false {
```

Comment:
This `if` is not needed. We don't need to log this - just let it go through to the return on line 217.